File: packages/core/src/checkpoint.ts
import type { ThreadStateSnapshot } from "./thread.js";

/**
 * Checkpointer interface: persist and restore workflow state for durable execution.
 * Implementations: in-memory, Redis, Postgres.
 */
export interface Checkpointer {
  put(snapshot: ThreadStateSnapshot): Promise<void>;
  get(threadId: string): Promise<ThreadStateSnapshot | null>;
  list(threadId: string): Promise<ThreadStateSnapshot[]>;
  delete(threadId: string, nodeId?: string): Promise<void>;
}

File: python/src/agent_framework/checkpoint.py
"""Checkpointer interface for durable execution."""

from abc import ABC, abstractmethod
from agent_framework.thread import ThreadStateSnapshot


class Checkpointer(ABC):
    """Persist and restore workflow state."""

    @abstractmethod
    async def put(self, snapshot: ThreadStateSnapshot) -> None:
        ...

    @abstractmethod
    async def get(self, thread_id: str) -> ThreadStateSnapshot | None:
        ...

    @abstractmethod
    async def list_snapshots(self, thread_id: str) -> list[ThreadStateSnapshot]:
        ...

    @abstractmethod
    async def delete(self, thread_id: str, node_id: str | None = None) -> None:
        ...
